Add flags for setting the colour properties of a video track.

chenchao1983 opened the pull request:
Dear Colleagues
Please kindly review this patch.

I have added a set of flags to enable configuring the colour properties for MKV container.
mkvinfo is also updated to show the colour properties.

mbunkus commented:
Thanks. I'll review it, but it'll take a couple of days as I'll be either working or out of town for most of the week.

chenchao1983 commented:
No problem. Thank you very much.

mbunkus commented on src/common/strings/parsing.h:
```diff
@@ -73,6 +74,46 @@ parse_number(StrT const &string,
+template<typename StructType, typename DataType>
+bool
+parse_property_to_struct(const std::string& s,
+             std::map<int64_t, StructType>& property_list) {
```
Please fix the alignment.

mbunkus commented on src/common/strings/parsing.h:
```diff
@@ -73,6 +74,46 @@ parse_number(StrT const &string,
+parse_property_to_struct(const std::string& s,
+             std::map<int64_t, StructType>& property_list) {
+  int64_t id = 0;
+  std::vector<std::string> v = split(s, ":");
```
Please use `auto` wherever possible, e.g. `auto v = split(s, ":");`

mbunkus commented on src/common/strings/parsing.h:
```diff
@@ -73,6 +74,46 @@ parse_number(StrT const &string,
+  std::vector<DataType> data;
+  for (int i = 0; i < StructType::num_properties; ++i) {
+    DataType value;
+    if(!parse_number(v[i], value))
```
Please leave one space between key words and the opening parenthesis.

mbunkus commented on src/common/strings/parsing.h:
```diff
@@ -73,6 +74,46 @@ parse_number(StrT const &string,
+parse_property_to_value(const std::string& s,
+                        std::map<int64_t, DataType>& property_list) {
+  int64_t id = 0;
+  std::vector<std::string> v = split(s, ":");
```
Please use `auto` wherever possible, e.g. `auto v = split(s, ":");`

mbunkus commented on src/common/strings/parsing.h:
```diff
@@ -73,6 +74,46 @@ parse_number(StrT const &string,
+    return false;
+
+  DataType value;
+  if(!parse_number(v[1], value))
```
Please leave one space between key words and the opening parenthesis.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    if (Is<KaxVideoRChromaX>(l6)) {
+      KaxVideoRChromaX &x = *static_cast<KaxVideoRChromaX *>(l6);
+      show_element(l6, 6, boost::format(Y("Red colour coordinate x: %1%")) % x.GetValue());
+      summary.push_back((boost::format(Y("Red colour coordinates x: %1%")) % x.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoRChromaY>(l6)) {
+      KaxVideoRChromaY &y = *static_cast<KaxVideoRChromaY *>(l6);
+      show_element(l6, 6, boost::format(Y("Red colour coordinate y: %1%")) % y.GetValue());
+      summary.push_back((boost::format(Y("Red colour coordinates y: %1%")) % y.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoGChromaX>(l6)) {
+      KaxVideoGChromaX &x = *static_cast<KaxVideoGChromaX *>(l6);
+      show_element(l6, 6, boost::format(Y("Green colour coordinate x: %1%")) % x.GetValue());
+      summary.push_back((boost::format(Y("Green colour coordinates x: %1%")) % x.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoGChromaY>(l6)) {
+      KaxVideoGChromaY &y = *static_cast<KaxVideoGChromaY *>(l6);
+      show_element(l6, 6, boost::format(Y("Green colour coordinate y: %1%")) % y.GetValue());
+      summary.push_back((boost::format(Y("Green colour coordinates y: %1%")) % y.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoBChromaX>(l6)) {
+      KaxVideoBChromaX &x = *static_cast<KaxVideoBChromaX *>(l6);
+      show_element(l6, 6, boost::format(Y("Blue colour coordinate x: %1%")) % x.GetValue());
+      summary.push_back((boost::format(Y("Blue colour coordinates x: %1%")) % x.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoBChromaY>(l6)) {
+      KaxVideoBChromaY &y = *static_cast<KaxVideoBChromaY *>(l6);
+      show_element(l6, 6, boost::format(Y("Blue colour coordinate y: %1%")) % y.GetValue());
+      summary.push_back((boost::format(Y("Blue colour coordinates y: %1%")) % y.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoWhitePointChromaX>(l6)) {
+      KaxVideoWhitePointChromaX &x = *static_cast<KaxVideoWhitePointChromaX *>(l6);
+      show_element(l6, 6, boost::format(Y("White colour coordinate x: %1%")) % x.GetValue());
+      summary.push_back((boost::format(Y("White colour coordinates x: %1%")) % x.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoWhitePointChromaY>(l6)) {
+      KaxVideoWhitePointChromaY &y = *static_cast<KaxVideoWhitePointChromaY *>(l6);
+      show_element(l6, 6, boost::format(Y("White colour coordinate y: %1%")) % y.GetValue());
+      summary.push_back((boost::format(Y("White colour coordinates y: %1%")) % y.GetValue()).str());
```
There's no reason to have two different strings to translate here. Please use the singular in both messages.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+  show_element(l5, 5, Y("Video colour mastering metadata"));
+  for (auto l6: *static_cast<EbmlMaster *>(l5)) {
+    if (Is<KaxVideoRChromaX>(l6)) {
+      KaxVideoRChromaX &x = *static_cast<KaxVideoRChromaX *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Red colour coordinates x: %1%")) % x.GetValue()).str());
+
+    } else if (Is<KaxVideoRChromaY>(l6)) {
+      KaxVideoRChromaY &y = *static_cast<KaxVideoRChromaY *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Red colour coordinates y: %1%")) % y.GetValue()).str());
+
+    } else if (Is<KaxVideoGChromaX>(l6)) {
+      KaxVideoGChromaX &x = *static_cast<KaxVideoGChromaX *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Green colour coordinates x: %1%")) % x.GetValue()).str());
+
+    } else if (Is<KaxVideoGChromaY>(l6)) {
+      KaxVideoGChromaY &y = *static_cast<KaxVideoGChromaY *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Green colour coordinates y: %1%")) % y.GetValue()).str());
+
+    } else if (Is<KaxVideoBChromaX>(l6)) {
+      KaxVideoBChromaX &x = *static_cast<KaxVideoBChromaX *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Blue colour coordinates x: %1%")) % x.GetValue()).str());
+
+    } else if (Is<KaxVideoBChromaY>(l6)) {
+      KaxVideoBChromaY &y = *static_cast<KaxVideoBChromaY *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Blue colour coordinates y: %1%")) % y.GetValue()).str());
+
+    } else if (Is<KaxVideoWhitePointChromaX>(l6)) {
+      KaxVideoWhitePointChromaX &x = *static_cast<KaxVideoWhitePointChromaX *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("White colour coordinates x: %1%")) % x.GetValue()).str());
+
+    } else if (Is<KaxVideoWhitePointChromaY>(l6)) {
+      KaxVideoWhitePointChromaY &y = *static_cast<KaxVideoWhitePointChromaY *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("White colour coordinates y: %1%")) % y.GetValue()).str());
+
+    } else if (Is<KaxVideoLuminanceMax>(l6)) {
+      KaxVideoLuminanceMax &luminance = *static_cast<KaxVideoLuminanceMax *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Max luminancne : %1%")) % luminance.GetValue()).str());
+
+    } else if (Is<KaxVideoLuminanceMin>(l6)) {
+      KaxVideoLuminanceMin &luminance = *static_cast<KaxVideoLuminanceMin *>(l6);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoLuminanceMax>(l6)) {
+      KaxVideoLuminanceMax &luminance = *static_cast<KaxVideoLuminanceMax *>(l6);
+      show_element(l6, 6, boost::format(Y("Max luminance : %1%")) % luminance.GetValue());
+      summary.push_back((boost::format(Y("Max luminancne : %1%")) % luminance.GetValue()).str());
```
There's a spelling error in the second usage of `luminance`. Please use the same string in both places.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoLuminanceMin>(l6)) {
+      KaxVideoLuminanceMin &luminance = *static_cast<KaxVideoLuminanceMin *>(l6);
+      show_element(l6, 6, boost::format(Y("Min luminance : %1%")) % luminance.GetValue());
+      summary.push_back((boost::format(Y("Min luminancne : %1%")) % luminance.GetValue()).str());
```
There's a spelling error in the second usage of `luminance`. Please use the same string in both places.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+  show_element(l4, 4, Y("Video colour"));
+  for (auto l5 : *static_cast<EbmlMaster *>(l4)) {
+    if (Is<KaxVideoColourMatrix>(l5)) {
+      KaxVideoColourMatrix &colour_matrix = *static_cast<KaxVideoColourMatrix *>(l5);
```
Please use auto wherever possible: `auto &x = *static_cast<…>(…)`

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+      summary.push_back((boost::format(Y("Colour matrix: %1%")) % colour_matrix.GetValue()).str());
+
+    } else if (Is<KaxVideoBitsPerChannel>(l5)) {
+      KaxVideoBitsPerChannel &bits_per_channel = *static_cast<KaxVideoBitsPerChannel *>(l5);
```
OK I'll stop here with `auto`; please just use it in those other places in this function. Thanks.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+    } else if (Is<KaxVideoChromaSitVert>(l5)) {
+      KaxVideoChromaSitVert &sit_vert = *static_cast<KaxVideoChromaSitVert *>(l5);
+      show_element(l5, 5, boost::format(Y("Vertical chroma sit : %1%")) % sit_vert.GetValue());
+      summary.push_back((boost::format(Y("Vertical chroma sit : %1%")) % sit_vert.GetValue()).str());
```
Please use the full word `siting` in these two messages just as in the horizontal case above.

mbunkus commented on src/info/mkvinfo.cpp:
```diff
@@ -570,6 +570,141 @@ handle_audio_track(EbmlStream *&es,
+
+    } else if (Is<KaxVideoLuminanceMax>(l6)) {
+      KaxVideoLuminanceMax &luminance = *static_cast<KaxVideoLuminanceMax *>(l6);
+      show_element(l6, 6, boost::format(Y("Max luminance : %1%")) % luminance.GetValue());
```
Here (and in a lot of the following messages) you're putting a space before the colon that separates the value from the message. Please remove that leading space here and in all following messages: `Max luminance: %1%`. Thanks.

mbunkus commented on src/input/r_matroska.cpp:
```diff
@@ -133,6 +133,41 @@ kax_track_t::handle_packetizer_pixel_cropping() {
+    ptzr_ptr->set_video_max_CLL(v_max_CLL, OPTION_SOURCE_CONTAINER);
+  if (v_max_FALL != -1)
+    ptzr_ptr->set_video_max_FALL(v_max_FALL, OPTION_SOURCE_CONTAINER);
+  if (v_chroma_coordinates.red_x != -1 || v_chroma_coordinates.red_y != -1 ||
```
Please use parenthesis around the parts separated by logical operators:

    if ((v_chroma_coordinates.red_x != -1) || (v_chroma_coordinates.red_y != -1) || …

Also, if you line break then please use my somewhat idiosyncratic style of moving the last logical operator to the start of the following line and aligning all the particular parts like this:

    if (   (v_chroma_coordinates.red_x   != -1) || (v_chroma_coordinates.red_y   != -1)
        || (v_chroma_coordinates.green_x != -1) || (v_chroma_coordinates.green_y != -1)
        || (v_chroma_coordinates.blue_x  != -1) || (v_chroma_coordinates.blue_y  != -1)) {

See how I've aligned the comparison operators, too? I'm a big fan of aligning such things because they make the code much easier to read for me.

mbunkus commented on src/input/r_matroska.cpp:
```diff
@@ -1014,6 +1049,38 @@ kax_reader_c::read_headers_track_video(kax_track_t *track,
+    track->v_colour_primaries = FindChildValue<KaxVideoColourPrimaries>(colour, -1);
+    track->v_max_CLL = FindChildValue<KaxVideoColourMaxCLL>(colour, -1);
+    track->v_max_FALL = FindChildValue<KaxVideoColourMaxFALL>(colour, -1);
+    track->v_max_FALL = FindChildValue<KaxVideoColourMaxFALL>(colour, -1);
```
Please align all the assignment operators in the same column in order to improve readability.

mbunkus commented on src/input/r_matroska.cpp:
```diff
@@ -1014,6 +1049,38 @@ kax_reader_c::read_headers_track_video(kax_track_t *track,
+    track->v_max_CLL = FindChildValue<KaxVideoColourMaxCLL>(colour, -1);
+    track->v_max_FALL = FindChildValue<KaxVideoColourMaxFALL>(colour, -1);
+    track->v_max_FALL = FindChildValue<KaxVideoColourMaxFALL>(colour, -1);
+    KaxVideoColourMasterMeta* colour_meta =  FindChild<KaxVideoColourMasterMeta>(*colour);
```
`auto colour_meta = FIndChild<…>…` :)

mbunkus commented on src/input/r_matroska.cpp:
```diff
@@ -1014,6 +1049,38 @@ kax_reader_c::read_headers_track_video(kax_track_t *track,
+      track->v_white_colour_coordinates.x = FindChildValue<KaxVideoWhitePointChromaX>(colour_meta, -1.0);
+      track->v_white_colour_coordinates.y = FindChildValue<KaxVideoWhitePointChromaX>(colour_meta, -1.0);
+      track->v_max_luminance = FindChildValue<KaxVideoLuminanceMax>(colour_meta, -1.0);
+      track->v_min_luminance = FindChildValue<KaxVideoLuminanceMin>(colour_meta, -1.0);
```
Align all of those on `=`, too, please.

mbunkus commented on src/merge/generic_packetizer.cpp:
```diff
@@ -574,6 +644,187 @@ generic_packetizer_c::set_video_pixel_cropping(int left,
+generic_packetizer_c::set_video_colour_matrix(int matrix_index,
+                                              option_source_e source) {
+  m_ti.m_colour_matrix.set(matrix_index, source);
+  if (m_track_entry && matrix_index >= 0 && matrix_index <= 10) {
```
Parenthesis around the different parts, please.

mbunkus commented on src/merge/generic_packetizer.cpp:
```diff
@@ -574,6 +644,187 @@ generic_packetizer_c::set_video_pixel_cropping(int left,
+  if (m_track_entry && matrix_index >= 0 && matrix_index <= 10) {
+    KaxTrackVideo &video = GetChild<KaxTrackVideo>(m_track_entry);
+    KaxVideoColour& color = GetChild<KaxVideoColour>(video);
+    GetChild<KaxVideoColourMatrix>(color).SetValue(m_ti.m_colour_matrix.get());
```
Use of `auto` and alignment on `=` here and all the following functions, please.

mbunkus commented on src/merge/generic_packetizer.cpp:
```diff
@@ -574,6 +644,187 @@ generic_packetizer_c::set_video_pixel_cropping(int left,
+generic_packetizer_c::set_video_white_colour_coordinates(
+    const white_colour_coordinates_t& coordinates, option_source_e source) {
+  m_ti.m_white_coordinates.set(
+      white_colour_coordinates_t(coordinates.x, coordinates.y), source);
```
Please don't line break here and not in such a way. I don't adhere to a low line length limit. Long lines are perfectly OK in most cases.

mbunkus commented on src/merge/generic_packetizer.cpp:
```diff
@@ -574,6 +644,187 @@ generic_packetizer_c::set_video_pixel_cropping(int left,
+    KaxVideoColour& color = GetChild<KaxVideoColour>(video);
+    GetChild<KaxVideoLuminanceMin>(color).SetValue(luminance);
+  }
+}
```
All of these functions need:

- alignment on `=`
- use of `auto`
- use of more parenthesis in `if`

mbunkus commented on src/merge/generic_packetizer.cpp:
```diff
@@ -727,6 +978,103 @@ generic_packetizer_c::set_headers() {
+        KaxVideoColour& colour = GetChild<KaxVideoColour>(video);
+        KaxVideoColourMasterMeta& master_meta = GetChild<KaxVideoColourMasterMeta>(colour);
+        GetChild<KaxVideoLuminanceMin>(master_meta).SetValue(luminance);
+      }
```
All of these functions need:

- alignment on `=`
- use of `auto`
- use of more parenthesis in `if`

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -261,6 +261,64 @@ set_usage() {
+                  "                           2: Unspecified, 3: Reserved, 4: FCC, 5: BT470BG,\n"
+                  "                           6: SMPTE 170M, 7: SMPTE 240M, 8: YCOCG, 9: BT2020 \n"
+                  "                           Non-constant Luminance, 10: BT2020 Constant Luminance)\n");
+  usage_text += Y("  --bits-per-channel <TID:n>\n"
```
This is misleading for users as they might suspect that such an option can be used with audio tracks. Please rename it to `--colour-bits-per-channel`.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -261,6 +261,64 @@ set_usage() {
+                  "                           every pixel not removed horizontally/vertically. Example: For\n"
+                  "                           video with 4:2:0 chroma subsampling, the parameter should be \n"
+                  "                           set to TID:1,1 .\n");
+  usage_text += Y("  --Cb-subsample <TID:hori,vert>\n"
```
Names of long options must be all-lowercase.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+   10: BT2020 Constant Luminance)
+*/
+static void
+parse_arg_colour_matrix(const std::string &s, track_info_c &ti) {
```
My idiosyncratic style of function definitions are to have each argument on its own line. Please fix this and all following functions to suit this style. Thanks.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+   The argument must have the form \c TID:n e.g. \c 0:8
+*/
+static void
+parse_arg_bits_per_channel(const std::string& s, track_info_c &ti) {
```
I put the reference operator next to the variable name, not next to the type (`std::string &s`). Please fix this and all the following occurrences.

mbunkus replied:
Oh, and please rename the function to `parse_arg_color_bits_per_channel` to suit the argument name I've proposed.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+static void
+parse_arg_max_frame_light(const std::string& s, track_info_c &ti) {
+  if (!parse_property_to_value<int>(s, ti.m_max_FALL_list))
+    mxerror(boost::format("Max content light parameter: not given in the form <TID>:n (argument was '%1%').") % s);
```
The error message does not match the function name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+static void
+parse_arg_chroma_coordinates(const std::string& s, track_info_c &ti) {
+  if (!parse_property_to_struct<chroma_coordinates_t, float>(s, ti.m_chroma_coordinates_list))
+    mxerror(boost::format("Cb subsampling parameter: not given in the form <TID>:hori,vert (argument was '%1%').") % s);
```
The error message does not match the function name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+static void
+parse_arg_white_coordinates(const std::string& s, track_info_c &ti) {
+  if (!parse_property_to_struct<white_colour_coordinates_t, float>(s, ti.m_white_coordinates_list))
+    mxerror(boost::format("Cb subsampling parameter: not given in the form <TID>:hori,vert (argument was '%1%').") % s);
```
The error message does not match the function name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+static void
+parse_arg_max_luminance(const std::string& s, track_info_c &ti) {
+  if (!parse_property_to_value<float>(s, ti.m_max_luminance_list))
+    mxerror(boost::format("Max content light parameter: not given in the form <TID>:n (argument was '%1%').") % s);
```
The error message does not match the function name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -727,6 +785,150 @@ parse_arg_cropping(const std::string &s,
+static void
+parse_arg_min_luminance(const std::string& s, track_info_c &ti) {
+  if (!parse_property_to_value<float>(s, ti.m_min_luminance_list))
+    mxerror(boost::format("Max content light parameter: not given in the form <TID>:n (argument was '%1%').") % s);
```
The error message does not match the function name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -2322,6 +2524,104 @@ parse_args(std::vector<std::string> args) {
+      if (no_next_arg)
+        mxerror(Y("'--bits-per-channel' lacks the parameter.\n"));
+
+      parse_arg_bits_per_channel(next_arg, *ti);
```
Change the parameter and function names as I've proposed earlier.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -2322,6 +2524,104 @@ parse_args(std::vector<std::string> args) {
+      if (no_next_arg)
+        mxerror(Y("'--Cb-subsample' lacks the parameter.\n"));
+
+      parse_arg_Cb_subsample(next_arg, *ti);
```
Change the parameter and function names to all-lowercase.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -2322,6 +2524,104 @@ parse_args(std::vector<std::string> args) {
+
+    } else if (this_arg == "--chromaticity-coordinates") {
+      if (no_next_arg)
+        mxerror(Y("'--chromaticity-color-coordinates' lacks the parameter.\n"));
```
The error message does not match the parameter name.

mbunkus commented on src/merge/mkvmerge.cpp:
```diff
@@ -2322,6 +2524,104 @@ parse_args(std::vector<std::string> args) {
+
+    } else if (this_arg == "--min-luminance") {
+      if (no_next_arg)
+        mxerror(Y("'--min-luminance' lacks the parameter.\n"));
```
All in all it would be better not to have different error messages for all those cases. This imposes a lot of work on the translators. Instead something like this should be used for all of these cases:

    mxerror(boost::format(Y("'%1%' lacks its argument.\n")) % "--min-luminance");

This exact string is already present and translated.

chenchao1983 replied:
I don't quite understand how translation is working. Could you educate me a bit on this? Why would mxerror(boost::format(Y("'%1%' lacks its argument.\n")) % "--min-luminance");
cost less computations than
mxerror(Y("'--min-luminance' lacks the parameter.\n"));

where is the macro Y being defined?

mbunkus replied:
The `Y(…)` macro marks its first parameter as to be translated. So in my version the string to translate is `%1%' lacks its argument.\n`.

Replacing only that one error message `'--min-luminance' lacks the parameter.\n` with the version I've proposed doesn't gain much, of course. What I meant was that you should replace all of those similar error messages with my version. So instead of using these (which are three distinct strings to translate):

    mxerror(Y("'--white-color-coordinates' lacks the parameter.\n"));
    …
    mxerror(Y("'--min-luminance' lacks the parameter.\n"));
    …
    mxerror(Y("'--max-luminance' lacks the parameter.\n"));

you should use these:

    mxerror(boost::format(Y("'%1%' lacks its argument.\n")) % "--white-color-coordinates");
    …
    mxerror(boost::format(Y("'%1%' lacks its argument.\n")) % "--max-luminance");
    …
    mxerror(boost::format(Y("'%1%' lacks its argument.\n")) % "--min-luminance");

In my version there's only a single string to translate. The three different parameter names are only inserted into the string after the translation putting much less stress on the translators.

You can read my version like this:

1. Translate the string `'%1%' lacks its argument.\n`.
2. Feed the translated string from 1. into the `boost::format()` formatter and give it `--white-color-coordinates` as its first and sole argument, meaning all occurrences of `%1%` in the (already translated) string will be replaced by `--white-color-coordinates`.
3. Take the formatted string and hand it over to the `mxerror()` function.

mbunkus commented on src/merge/track_info.h:
```diff
@@ -98,6 +98,126 @@ struct pixel_crop_t {
+  const static int num_properties = 2;
+};
+
+struct Cb_subsample_t {
```
Please name all types in all-lowercase. Exceptions are only made for types used in the GUIs to match Qt's naming conventions more closely, but not in mkvmerge itself.

mbunkus commented on src/merge/track_info.h:
```diff
@@ -98,6 +98,126 @@ struct pixel_crop_t {
+  }
+  chroma_coordinates_t(float p_red_x, float p_red_y,
+                    float p_green_x, float p_green_y,
+                    float p_blue_x, float p_blue_y)
```
Please fix the alignment.

mbunkus commented on src/merge/track_info.h:
```diff
@@ -176,6 +296,48 @@ class track_info_c {
+  option_with_source_c<chroma_subsample_t> m_chroma_subsample; // For this very track
+
+  std::map<int64_t, Cb_subsample_t> m_Cb_subsample_list; // As given on the command line
+  option_with_source_c<Cb_subsample_t> m_Cb_subsample; // For this very track
```
Please fix the variable and type names to all-lowercase. Thanks.

mbunkus commented on src/merge/track_info.h:
```diff
@@ -176,6 +296,48 @@ class track_info_c {
+  option_with_source_c<int> m_colour_primaries; // For this very track
+
+  std::map<int64_t, int> m_max_CLL_list; // As given on the command line
+  option_with_source_c<int> m_max_CLL; // For this very track
```
Again, all-lowercase, please.

mbunkus commented on src/merge/track_info.h:
```diff
@@ -176,6 +296,48 @@ class track_info_c {
+  option_with_source_c<int> m_max_CLL; // For this very track
+
+  std::map<int64_t, int> m_max_FALL_list; // As given on the command line
+  option_with_source_c<int> m_max_FALL; // For this very track
```
Once more all-lowercase.

mbunkus commented:
Alright, I'm through with a first review. Thanks again for the work!

Yes, I am anal about the styling used, but that's simply to make my life as the maintainer of all the code easier. I'm not pointing all those things out just to annoy you.

I haven't given the patch a spin yet due to the number of issues I've marked and due to the fact that it currently doesn't apply cleanly. Please re-base in addition to fixing the things I've mentioned.

Apart from all your work (for which I'm grateful) there are several things to do:

1. add documentation in mkvmerge's man page (`doc/man/mkvmerge.xml`) for all of the new options
2. bump the configure check to check for a newer libMatroska version (this can only be done once I've released the next libMatroska version, of course) in order to ensure the newly added elements are truly available
3. add an entry to the `ChangeLog` about the new options
4. add you to the `AUTHORS` file with an appropriate entry
5. add controls in the GUI in the "merge" tool
6. add support in `mkvpropedit` for all those elements
7. add support in the GUI's header editor for all those elements

All of that is quite a lot of work. I'll definitely do 2. myself once the new version is out. 3. and 4. are easy enough to do; you can do it yourself if you want (preferably as I don't know your real name yet), otherwise I'll do it myself.

Do you have plans to work on 1.? I'd appreciate it, but I'd do it myself, too.

The points 5. through 7. aren't important for merging this particular pull request. I've only mentioned them for the sake of completeness.

chenchao1983 commented:
Thank you very much for reviewing the code. I have addressed all comments. Please have another look.
I will 1,3,4 soon and finish 5,6,7 in the near future.
When are you going to roll out the next release?

best,
Chao Chen

mbunkus commented:
Thanks. Can you please squash the two commits and force-push? A single commit is easier to review, and at this stage we don't need the history. Thanks.

The next release is currently scheduled in four weeks.

chenchao1983 commented:
No problem. I have combined the two commits together. Thanks.

mbunkus commented:
Thanks. I'll review it tomorrow or on Sunday.

mbunkus commented:
Due to illness I haven't gotten around to reviewing it yet. It is on my TODO list; please be patient. Thanks.

chenchao1983 commented:
Sure, no problem.

chenchao1983 commented:
Hi, Moritz
Could you help review this request and make a release base on it?
Thank you very much.
best,
Chao Chen

mbunkus commented:
I will when I find the time of which I currently don't have a lot. As this also requires a new release of libMatroska it has impact on other users of it, too. I definitely haven't forgotten about it.

chenchao1983 commented:
No, problem. Will do it soon.



Chao Chen

On Mon, Jun 6, 2016 at 10:23 AM, Moritz Bunkus <notifications@github.com>
wrote:

> Again thanks for your work. I've reviewed the pull request. During that
> review I've implemented a number of cosmetic changes (alignment,
> indentation, placement of &, placement of const etc.). I've uploaded the
> diff here
> <https://mkvtoolnix.download/misc/add_flags_for_video_colour-cosmetics-1.diff>.
> Please apply it and rebase the pull request to the current master branch
> (only one small conflict in src/common/strings/parsing.h if I remember
> correctly).
>
> I have one thing I'd like you to fix/change before I merge it and take
> care of the rest: the help texts output by mkvmerge. Two things:
>
>    1. You've added a lot of information about the valid values and what
>    they mean. I tend not to do that in the --help texts but in the man
>    pages. The --help output is very, very long already, and its intention
>    is to give a terse overview of the valid syntax – not of all valid values
>    (compare to e.g. --stereo-mode). So please change the texts to be more
>    in line with the rest of the explanations to the man page,
>    doc/man/mkvmerge.xml.
>    2. While you're at it please restrict the length of all those lines to
>    98 columns. This results in the actual output not being longer than 79
>    columns fitting into the terminal sizes that are usually used by default.
>
> That should really be the last I'm going to ask of you before accepting
> it. Thanks!
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mbunkus/mkvtoolnix/pull/1675#issuecomment-224027012>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AGvyQ2D59cApiKcRJ4vTlAEua7e-UTwPks5qJFetgaJpZM4IPmLs>
> .
>

chenchao1983 commented:
Dear Colleague
I have made the changes accordingly. Please have a look.

mbunkus commented:
Thanks. Just as a heads-up: I'll be out of town for ~two weeks. Afterwards I will make time to finally merge your pull requests.

mbunkus commented:
I've spent some time today merging this commit and fixing a lot of issues with the formatting. Unfortunately you misunderstood my last request, or you misconfigured your editor – and the result was that `mkvmerge.cpp` was reformatted almost completely. I've undone all that and folded those changes into your commit. So you don't have to do anything about this.

I will most likely take care of your other pull request (documentation, ChangeLog, AUTHORS) tomorrow. Afterwards I'll close both pull requests.

Progress! :)

mbunkus commented:
Cherry-picked and fixed style issues in 8d8d6d78a0be569154233cb50e647831cd39a68b. Thanks.

New [pre-builds for Windows](https://mkvtoolnix.download/windows/pre/) are available. Build numbers 01251 and higher contain both this and the documentation pull request.
